### PR TITLE
Add support for CAN RxCallbacks

### PR DESCRIPTION
**Pull request by HenrikSolver**
This patch adds support for adding callbacks to a can class that will be called when a message is received.

### CODE REVIEW

File: stmhal/can.c [hunk 1]
```diff
@@ -525,6 +556,52 @@ STATIC mp_obj_t pyb_can_setfilter(mp_uint_t n_args, const mp_obj_t *pos_args, mp
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_KW(pyb_can_setfilter_obj, 1, pyb_can_setfilter);
 
+STATIC mp_obj_t pyb_can_rxcallback(mp_obj_t self_in, mp_obj_t fifo_in, mp_obj_t callback_in) {
+    pyb_can_obj_t *self = self_in;
+    mp_int_t fifo = mp_obj_get_int(fifo_in);
+
+
+    if (callback_in == mp_const_none){
```
Comment (pfalcon) [inline stmhal/can.c]: Here, and in other places, missing space between ')' & '{'.

File: stmhal/can.c [hunk 2]
```diff
@@ -525,6 +556,52 @@ STATIC mp_obj_t pyb_can_setfilter(mp_uint_t n_args, const mp_obj_t *pos_args, mp
...
+            __HAL_CAN_DISABLE_IT(&self->can, CAN_IT_FMP0);
+            self->rxcallback0 = mp_const_none;
+        } else if (fifo == 1){
+            __HAL_CAN_DISABLE_IT(&self->can, CAN_IT_FMP1);
+            self->rxcallback1 = mp_const_none;
+        }
+    } else if (mp_obj_is_callable(callback_in)) {
+        if (self->can_id == PYB_CAN_1){
+            if (fifo == 0) {
+                self->rxcallback0 = callback_in;
```
Comment (pfalcon) [inline stmhal/can.c]: This code snippet is repeated 4 times. Would it make sense to set values based on conditions and then call functions once?

Comment (PappaPeppar) [inline stmhal/can.c]: I guess it is a matter of style. Personally I think this way of writing code is a bit easier to read an understand, but I will not make any noise if you want me to change it before accepting the code.

Comment (pfalcon) [inline stmhal/can.c]: > I guess it is a matter of style. Personally I think this way of writing code is a bit easier to read an understand, but I will not make any noise if you want me to change it before accepting the code.

Well, I'd say that writing code such way hides the fact that what those if's do is actually transcoding values, but operations on ready values are the same. Suppose you'd need to change these operations (add another one, or rename a function) - these would need to be done in 4 places (what if there were more?). In more complex cases someone else doing a refactor could even miss to update a branch. Finally, doing multiple calls will almost certainly costs more bytes - and it seems that to completely wrap CAN there will be need quite a few of them, and there're lot of other hardware to wrap yet...

Anyway, it's up to @dpgeorge to consider how that's important.

Anyway, I started to look at your patch because for esp8266 port there soon would be need for callbacks, and that's where there're lot of boilerplate code and what will need to be refactored on uPy level...

**Comment by dpgeorge** [comment]
@PappaPeppar thanks!  There are a few things I'll clean up and fix (like putting pyb_can_obj_all in root pointer section), but main thing is to discuss how to call the callback function.

Things like ExtInt and Timer only need one callback, when the event happens.  Things like UART can have many callbacks: when data is received, when tx buffer is half empty, when tx buffer is completely empty.  DAC can have callback when buffer is half empty, SPI can have callback when DMA transfer is finished.  Etc.

Basically a callback is an interrupt request.

So how do we want to name the functions that set up the callbacks?  Do we want to use individual names, like callback, rxcallback, txemptycallback, dmacallback, etc; or do we want one name (callback) with an argument to set the callback type?  Eg can.callback('rx', lambda:...)?  Could use short codes like 'txhe' for tx buf half empty.  This would scale a bit better than adding many functions.

Any other ideas or comments?

**Comment by PappaPeppar** [comment]
First a short recap on the hardware capabilities in the area. The hardware supports three types of interrupts. Receive, Transmit and Error/Status information. 
__There are three Rx interrupts for each FIFO.__
* FIFO not empty: This one we need and this is what's implemented in this pull request.
* FIFO Full.
* FIFO Overrun: At first I thought that this wasn't needed, but we would definitely want to know if packets are lost due to overruns.

__There is one Tx interrupt for each bus controller.__
It is triggered when a message is sent, but if there are more than one message sent between two interrupts (that can happen, there are three buffers that can hold messages to be transmitted) it is not trivial to determine which message that actually caused the interrupt. So I think that the value of a Tx interrupt might be limited. But I have code ready, I only have to add docs and tests.

__There is one Error/Status interrupt for each bus controller.__
The can controller collects statistics about how many faults it sees on the bus. If these numbers go over specific thresholds it generates a interrupt, this is defined in the CAN spec. This interrupt can also be triggered by specific events/state changes in the hw. I have not started on this yet.

I would like to keep these events as separate callbacks.
* One Rx callback for each FIFO with a parameter that indicates the reason as above. The motivation for one callback for each FIFO is that if you make a design decision to receive the messages in different FIFOS you probably have reason. Why should then the python bindings merge the code paths? Another thing I am thinking about is that we might want to set different priorities on the interrupts, in that case the callback code must be reentrant.
* Tx callbacks should be optional. In many cases there is no need to for a Tx callback in which case a empty interrupt is a waste of cycles.

Regarding naming: rxcallback and txcallback are obvious to me, but I guess other might have other opinions. I have not thought much about a name for the third callback, errcallback might be a candidate.

**Comment by dpgeorge** [comment]
Some style changes, put pyb_can_obj_all in global state structure, and merged in f80f1a70777dd84171af02200b77033a08fba28a.
